content/raw/chat/chat_input.md

```diff
@@ -0,0 +1,127 @@
+2. User presses SPACE or ESC
+3. System closes the suggestions menu and returns focus to the text entry box.
+4. System places the text caret at the end of whatever the user has typed, including whatever the user typed while navigating and/or closing the suggestions menu.   
+USE CASE ENDS
+
+### Emojis
+
+TODO
+
+### Code Tags
+
+TODO
+
+### Text Formatter
+
+TODO
+
+## Functional Requirements
+
+**preconditions**
+
+- The user MUST be in a chat
+
+### Mentions
+
+User types a message and when the character `@` is typed it opens a menu with a list of suggestions, the user can then select a suggestion to autocomplete it
+
+- Typing `@` MUST open a suggestions menu
+- Typing after the `@` character will filter the suggestions to match what is being suggested
+- Initially the ordering of the suggestions SHOULD follow an alphabetical order
+- When a user starts the search the ordering of the suggestions menu MUST follow the following criteria in order
+    1. Results that match the search in the beginning of the username
+    2. Results that closely match the query
+    3. Alphabetical
+- The suggestion menu MUST have a selector
+- Pressing the key `Enter` or the key `Tab` MUST autocomplete with the currently selected mention
+
+### Emojis
+
+User types a message and once the character `:` is typed the next character type will either autoreplace the text with an emoji or open a suggestions menu
+
+- Typing `:` followed by a known character that forms an emoji MUST auto replace it with an emoji for example typing `:)` will replace it with 🙂
+- Typing `:` followed by one or more alphabetical characters MUST open a suggestions menu
+- When a user starts the search the ordering of the suggestions menu MUST follow the following criteria in order
+    1. Results that match the search in the beginning of the emoji name
+    2. Results that closely match the query
+    3. Alphabetical
+- The suggestion menu MUST have a selector
+- Pressing the key `Enter` or the key `Tab` MUST autocomplete with the currently selected emoji
+
+### Code Tags
+
+TODO
+
+### Text Formatter
+
+User types a message and afterwards can select part of his message and apply styles to it such as bold, italic, strikethrough or code tag. The user can also use markdown while typing the tag to apply these styles (e.g `**bold**``)
+
+- Selecting text with the cursor MUST display the formatting menu
+- Selecting text with the keyboard (shift + arrow keys) SHOULD display the formatting menu
+- Selecting one of the options in the formatting menu MUST change the selected text to reflect the markdown that applies that style
+- When an option in the formatting menu is selected the formatting menu SHOULD remain open
+- When text is selected with the keyboard, key combinations MUST apply that style and change the selected text to reflect the markdown that applies that style
+    - CTRL+B (windows/linux) CMD+B (mac) - bold
+    - CTRL+I (windows/linux) CMD+I (mac) - italic
+
```

Comment:
Editing of styles must work as follows:
- highlighting styled text and pressing the applied style dismisses the style
- removing special characters that define style dismisses the style